fendermint/vm/interpreter/src/fvm/upgrades.rs

```diff
@@ -1,54 +1,94 @@
 // Copyright 2022-2024 Protocol Labs
 // SPDX-License-Identifier: Apache-2.0, MIT
 
-use std::collections::BTreeMap;
+use std::{collections::BTreeMap, path::Path};
 
-use anyhow::bail;
-use fendermint_vm_core::chainid;
+use anyhow::{bail, Context};
+use fendermint_vm_message::ipc::UpgradeInfo;
 use fvm_ipld_blockstore::Blockstore;
-use fvm_shared::chainid::ChainID;
+use serde::{Deserialize, Serialize};
 use std::collections::btree_map::Entry::{Occupied, Vacant};
 
-use super::state::{snapshot::BlockHeight, FvmExecState};
+use super::state::FvmExecState;
+use fvm_shared::clock::ChainEpoch;
 
-#[derive(PartialEq, Eq, Clone)]
-struct UpgradeKey(ChainID, BlockHeight);
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct UpgradeSchedule {
+    // schedule of upgrades to be executed at given height
+    schedule: BTreeMap<ChainEpoch, UpgradeInfo>,
+}
 
```

Comment:
What I see missing here is a file system watcher. The JSON is only read when the node starts. We want it to stop when a new upgrade is announced, and the height arrives with the old version still running. If it doesn't refresh the file, then it won't know. 

I reckon it could keep a reference to the file and start a background loop that checks it every now and then, then loads new items into the `schedule`. 

This is also important because without a governance module, the upgrade schedule somehow has to make it to all full nodes, which is why I think the visor should be able to download it and overwrite the file on the disk. Not sure how to make sure that doesn't happen for the guy who is adding upgrades by hand.